src/mlpack/core/data/tokenizers/BERT_tokenizer.hpp

```diff
@@ -0,0 +1,257 @@
+/**
+ * @file core/data/tokenizers/BERT_tokenizer.hpp
+ * @author Ayush Singh
+ *
+ * Definition of BERT Tokenizer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_BERT_TOKENIZER_HPP
+#define MLPACK_CORE_DATA_BERT_TOKENIZER_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core.hpp>
+#include <vector>
+#include <unordered_map>
+#include <boost/algorithm/string.hpp>
+#include <string>
+
+namespace mlpack {
+namespace data {
+
+/**
+ * To use a pre-trained BERT model, we need to convert the input data into an
+ * appropriate format so that each sentence can be sent to the pre-trained model 
+ * to obtain the corresponding embedding.
+ * BERT Tokenizer breaks down the input string into tokens which can be easily 
+ * encoded by the BERT model.
+ * 
+ * For more information, see the following.
+ *
+ * @code
+ * @inproceedings{ACL2019,
+ *   title  = {BERT: Pre-training of Deep Bidirectional Transformers 
+ *             for Language Understanding},
+ *   author = {Jacob Devlin, Ming-Wei Chang, Kenton Lee, Kristina Toutanova},
+ *   year   = {2019},
+ *   url    = {https://arxiv.org/abs/1810.04805}
+ * }
+ * @endcode
+ */
+
+// All the characters which cause break in tex in a string.
+const std::wstring stripChar = L" \t\n\r\v\f";
+
+// The datatype in which the vocab file is stored. Used for mapping token with its id.
+using Vocab = std::unordered_map<std::wstring, size_t>;
+
+// The datatype in which the vocab file is stored. Used for mapping id with its token.
+using InvVocab = std::unordered_map<size_t, std::wstring>;
+
+/**
+ * This class is used for basic tokenization purposes.
+ * Tasks like punctuation splitting, accent splitting, 
+ * whitespace  and lower casing.
+ */
+class BasicTokenizer {
+ public:
+  /**
+   * Enable a check condition for conversion to lower case.
+   * 
+   * @param doLowerCase Pass true if you want to process text after 
+   * converting all characters to lower case.
+   */
+  BasicTokenizer(bool doLowerCase);
+
+  /**
+   * Execute basic tokenization by calling all the functions of this
+   * class in a sequential manner.
+   *
+   * @param text The input text.
+   */
+  std::vector<std::wstring> tokenize(const std::string& text) const;
```

Comment:
The tokenizer used a `std::string` as input and gives output as `vector<std::string>`. All the processing inside is done on `std::wstring` format. I earlier implemented templates, but realized that if some other data format shall be given, it would anyway result in an error and initializing the data type can make the code easier to understand.
If it is necessary, then I would again switch to tokenizer approach.

Reply:
But what I'm not understanding here is what the purpose of converting from `std::string` to `std::wstring` is.  If the user needs to pass a `std::string`, then there is no advantage to convert to a `std::wstring`.